cmd/kola/nfs.go

```diff
@@ -0,0 +1,205 @@
+	"time"
+
+	"github.com/coreos/mantle/cli"
+	"github.com/coreos/mantle/platform"
+
+	"github.com/coreos/coreos-cloudinit/config"
+)
+
+func init() {
+	cli.Register(cmdNfs)
+}
+
+var cmdNfs = &cli.Command{
+	Run:     runNfs,
+	Name:    "nfs",
+	Summary: "nfs client/server test (requires root)",
+	Usage:   "",
+	Description: `
+Test that the kernel NFS server and client work within CoreOS.
+`}
+
+func runNfs(args []string) int {
+	if len(args) != 0 {
+		log.Printf("No args accepted")
+		return 2
+	}
+
+	c, err := platform.NewQemuCluster()
+	if err != nil {
+		log.Printf("NewQemuCluster: %s", err)
+		return 2
+	}
+
+	defer func() {
+		if err := c.Destroy(); err != nil {
+			log.Printf("Cluster.Destroy: %s", err)
+			return
+		}
+	}()
+
+	/* server machine */
+	c1 := config.CloudConfig{
+		CoreOS: config.CoreOS{
+			Units: []config.Unit{
+				config.Unit{
+					Name:    "rpcbind.service",
+					Command: "start",
+				},
+				config.Unit{
+					Name:    "rpc-statd.service",
+					Command: "start",
+				},
+				config.Unit{
+					Name:    "rpc-mountd.service",
+					Command: "start",
+				},
+				config.Unit{
+					Name:    "nfsd.service",
+					Command: "start",
+				},
+			},
+		},
+		WriteFiles: []config.File{
+			config.File{
+				Content: "/tmp	*(ro,insecure,all_squash,no_subtree_check,fsid=0)",
+				Path: "/etc/exports",
+			},
+		},
+		Hostname: "nfs1",
+	}
+
+	m1, err := c.NewMachine(c1.String())
+	if err != nil {
+		log.Printf("Cluster.NewMachine: %s", err)
+		return 2
+	}
+
+	defer func() {
+		if err := m1.Destroy(); err != nil {
+			log.Printf("Machine.Destroy: %s", err)
+			return
+		}
+	}()
+
+	log.Printf("NFS server booted.")
+
+	/* poke a file in /tmp */
+	rnd := make([]byte, 8)
+	_, err = rand.Read(rnd)
+	if err != nil {
+		log.Printf("rand.Read: %s", err)
+		return 2
+	}
+
+	_, err = m1.SSH(fmt.Sprintf("sudo touch /tmp/%X", rnd))
+	if err != nil {
+		log.Printf("Machine.SSH: %s", err)
+		return 2
+	}
+
+	log.Printf("Test file created on server.")
+
+	/* client machine */
+
+	nfstmpl := `[Unit]
+Description=NFS Client
+After=network-online.target
+Requires=network-online.target
+After=rpc-statd.service
+Requires=rpc-statd.service
+
+[Mount]
+What=%s:/tmp
+Where=/mnt
+Type=nfs
+Options=defaults,noexec
+`
+
+	c2 := config.CloudConfig{
+		CoreOS: config.CoreOS{
+			Units: []config.Unit{
+				config.Unit{
+					Name:    "rpc-statd.service",
+					Command: "start",
+				},
+				config.Unit{
+					Name:    "mnt.mount",
+					Command: "start",
+					Content: fmt.Sprintf(nfstmpl, m1.IP()),
+				},
+			},
+		},
+		Hostname: "nfs2",
+	}
+
+	m2, err := c.NewMachine(c2.String())
+	if err != nil {
+		log.Printf("Cluster.NewMachine: %s", err)
+		return 2
+	}
+
+	defer func() {
+		if err := m2.Destroy(); err != nil {
+			log.Printf("Machine.Destroy: %s", err)
+			return
+		}
+	}()
+
+	log.Printf("NFS client booted.")
+
+	var lsmnt []byte
+
+	log.Printf("Waiting for NFS mount on client...")
+
+	/* there's probably a better wait to check the mount */
```

Comment:
At some point I would like to add support to go-systemd to interact with the dbus API tunneled over SSH (systemctl supports this in combination with systemd-stdio-bridge). But that is a fair amount of work and not a priority. So busy waits will work just dandy for now. Later we can watch the status of units instead.